pylink/library.py

```diff
@@ -423,3 +454,91 @@ def dll(self):
           A ``ctypes`` DLL instance if one was loaded, otherwise ``None``.
         """
         return self._lib
+
+
+class JLinkarmDlInfo:
+    """Helper used to retrieve the absolute path of the JLink library (aka DLL)
+    based on its so name (Linux).
+
+    This is used on Linux, where ctypes.util.find_library() will not return
+    the library full file path, but only the file name (aka so name).
+    We'll then rely on the native dlinfo() API to retrieve the library absolute
+    file path.
+
+    For e.g.:
+    - LD_LIBRARY_PATH=/mnt/platform/segger/JLink
+    - JLinkarmDllInfo('jlinkarm').path -> /mnt/platform/segger/JLink/libjlinkarm.so.7
+
+    Technically this is a wrapper to the `struct dlinfo` data answered
+    by the (Linux) dynamic linker.
+
+    The dlinfo() dance implementation is adapted from @cloudflightio,
+    https://github.com/cloudflightio/python-dlinfo.
+    """
+
+    # Request to obtain a pointer to the link_map structure corresponding
+    # to a given handle (Linux).
+    # See: man dlinfo(3)
+    RTLD_DI_LINKMAP = 2
+
+    # dlinfo(3): struct link_map, where ld_name will be the file path.
+    class LinkMap(ctypes.Structure):
+        """
+        Represents a C struct link_map (Linux).
+        See: man dlinfo(3)
+        """
+        _fields_ = [
+            ('l_addr', ctypes.c_void_p),
+            ('l_name', ctypes.c_char_p),
+            ('l_ld', ctypes.c_void_p),
+            ('l_next', ctypes.c_void_p),
+            ('l_prev', ctypes.c_void_p),
+        ]
+
+    def __init__(self, jlinkarm_soname):
+        """
+        Initialize the dlinfo struct for given soname.
+
+        The 'path' property will then hold the absolute path
+        of the corresponding shared library.
+
+        Args:
+        - jlinkarm_soname: The soname returned by find_library(),
+          for e.g. 'libjlinkarm.so.7'.
+        """
+        self._dll_path = None
+
+        tmp_cdll_jlink = ctypes.cdll.LoadLibrary(jlinkarm_soname)
+
+        # dlinfo() dance to retrieve the full library filepath.
+        #
+        dl_soname = ctypes_util.find_library('dl')
+        if dl_soname is not None:
+            tmp_cdll_dl = ctypes.cdll.LoadLibrary(dl_soname)
+            dlinfo = tmp_cdll_dl.dlinfo
+            dlinfo.argtypes = ctypes.c_void_p, ctypes.c_int, ctypes.c_void_p
+            dlinfo.restype = ctypes.c_int
+
+            linkmap = ctypes.c_void_p()
+            if dlinfo(tmp_cdll_jlink._handle, JLinkarmDlInfo.RTLD_DI_LINKMAP, ctypes.byref(linkmap)) == 0:
+                linkmap = ctypes.cast(linkmap, ctypes.POINTER(JLinkarmDlInfo.LinkMap))
+                self._dll_path = linkmap.contents.l_name.decode(sys.getdefaultencoding())
```

Comment:
Is this call guaranteed to succeed?

Reply:
Once the call to `dlinfo()` has succeeded:

- casting the *pointer* to `JLinkarmDlInfo.LinkMap`, and *dereferencing* it to access the pointed to `struct link_map` members, is guaranteed to succeed until breaking changes to the `RTLD_DI_LINKMAP` request API/ABI (which is unlikely to happen any time soon)
- `l_name`, the library absolute path, is of type `char *` and `decode()` should also succeed just fine (I can't see how the encoding used by `dlinfo()` when populating the struct could be incompatible with the system default locale)

Hence, while these two lines of code could actually fail, IMHO we should not prevent any exception to propagate:
- in case of incompatible encodings, the exception will warn the user about a probably broken system configuration
- in case of incompatible changes to the `struct link_map`'s fields layout, we're not even sure to *benefit* from an exception, since we'll forcibly cast some memory content to a `char *`: here `decode()` may actually fail, but may also *successfully* return some random string, and the error won't show until `LoadLibrary()` is called with this eccentric file path

Would you prefer we (silently) catch any possible exception here ?

Reply:
SGTM. I don't think we need to catch an exception here.